In [257]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn import datasets
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
%config InLineBackend.figure_format = 'svg'
from sklearn.metrics import classification_report

## 1. congressional_voting_dataset

### Предобработка данных

In [258]:
cvd = pd.read_csv("C:/Users/Даниил/Desktop/ОМО/congressional_voting_dataset.csv")
#cvd

In [259]:
# Получаем список значении
x = cvd.iloc [:, :-1].values

# Получаем список первичного параметра
y = cvd['political_party'].values

In [260]:
x

array([['n', 'y', 'n', ..., 'y', 'n', 'y'],
       ['n', 'y', 'n', ..., 'y', 'n', '?'],
       ['?', 'y', 'y', ..., 'y', 'n', 'n'],
       ...,
       ['n', '?', 'n', ..., 'y', 'n', 'y'],
       ['n', 'n', 'n', ..., 'y', 'n', 'y'],
       ['n', 'y', 'n', ..., 'y', '?', 'n']], dtype=object)

In [261]:
# Перевод основного показателя в численный вид
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
le.fit(y)
y = le.transform(y)

# Приводим названия в классы для отображения(?)
cvd_pred_names = le.classes_
cvd_pred_names


array(['democrat', 'republican'], dtype=object)

In [262]:
# Переводим значения в численный вид
x = cvd[cvd.columns[:-1]].apply(le.fit_transform)
x



handicapped_infants  water_project_cost_sharing  \
0                      1                           2   
1                      1                           2   
2                      0                           2   
3                      1                           2   
4                      2                           2   
..                   ...                         ...   
430                    1                           1   
431                    1                           1   
432                    1                           0   
433                    1                           1   
434                    1                           2   

     adoption_of_the_budget_resolution  physician_fee_freeze  el_salvador_aid  \
0                                    1                     2                2   
1                                    1                     2                2   
2                                    2                     0                2   
3                                    2                     1                0   
4                                    2                     1                2   
..                                 ...                   ...              ...   
430                                  2                     2                2   
431                                  2                     1                1   
432                                  1                     2                2   
433                                  1                     2                2   
434                                  1                     2                2   

     religious_groups_in_schools  anti_satellite_test_ban  \
0                              2                        1   
1                              2                        1   
2                              2                        1   
3                              2                        1   
4                              2                        1   
..                           ...                      ...   
430                            2                        1   
431                            1                        2   
432                            2                        1   
433                            2                        0   
434                            2                        1   

     aid_to_nicaraguan_contras  mx_missile  immigration  \
0                            1           1            2   
1                            1           1            1   
2                            1           1            1   
3                            1           1            1   
4                            1           1            1   
..                         ...         ...          ...   
430                          1           2            2   
431                          2           2            2   
432                          1           1            1   
433                          0           0            0   
434                          1           1            2   

     synfuels_corporation_cutback  education_spending  superfund_right_to_sue  \
0                               0                   2                       2   
1                               1                   2                       2   
2                               2                   1                       2   
3                               2                   1                       2   
4                               2                   0                       2   
..                            ...                 ...                     ...   
430                             1                   2                       2   
431                             1                   1                       1   
432                             2                   2                       2   
433                             1                   2                       2   
434                       

In [263]:
from sklearn.model_selection import train_test_split

# Разделяем датасет на обучающие признаки, тестовые признаки, целевые обучающие, целевые тестовые
x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.3, random_state=0)

In [264]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
sc.fit(x_train)
x_train_std = sc.transform(x_train)
x_test_std = sc.transform(x_test)

In [265]:
# Проверка перевода параметров в численный вид
x_train[:5], x_train_std[:5]

(     handicapped_infants  water_project_cost_sharing  \
 434                    1                           2   
 246                    1                           1   
 393                    0                           0   
 106                    1                           2   
 274                    2                           1   
 
      adoption_of_the_budget_resolution  physician_fee_freeze  el_salvador_aid  \
 434                                  1                     2                2   
 246                                  2                     1                2   
 393                                  0                     0                1   
 106                                  1                     2                2   
 274                                  1                     2                2   
 
      religious_groups_in_schools  anti_satellite_test_ban  \
 434                            2                        1   
 246                            1     

### Логистическая регрессия

In [266]:
from sklearn.linear_model import LogisticRegression

# Создаем логистическую регрессию и обучаем ее (fit) 
lr = LogisticRegression(C=100.0, random_state = 1)
lr.fit(x_train_std, y_train)

LogisticRegression(C=100.0, random_state=1)

In [267]:
lr.predict_proba(x_test_std[:3, :])

array([[9.99853179e-01, 1.46820541e-04],
       [9.99822730e-01, 1.77270130e-04],
       [9.99998958e-01, 1.04218607e-06]])

In [268]:
# Выдает значения предсказания с наибольшей вероятностью
lr.predict_proba(x_test_std[:10, :]).argmax(axis=1)

array([0, 0, 0, 1, 1, 0, 0, 1, 0, 1], dtype=int64)

In [269]:
# Делаем тестовое предсказание
lr.predict(x_test_std[0, :].reshape(1, -1)) 

array([0])

In [270]:
x_test_std[0, :]

array([-0.70749995, -0.44367825,  0.78211795, -0.69484868, -0.78060166,
        0.77727506,  0.81602509,  0.82251506,  0.95508403,  0.92776603,
        1.3190563 , -0.53983225,  1.01849365,  0.82961654, -0.59318861,
        0.69376162])

In [271]:
x_test_std[0, :].reshape(1, -1)

array([[-0.70749995, -0.44367825,  0.78211795, -0.69484868, -0.78060166,
         0.77727506,  0.81602509,  0.82251506,  0.95508403,  0.92776603,
         1.3190563 , -0.53983225,  1.01849365,  0.82961654, -0.59318861,
         0.69376162]])

In [272]:
y_pred = lr.predict(x_test_std)

In [273]:
print(classification_report(y_test, y_pred, target_names=cvd_pred_names))

              precision    recall  f1-score   support

    democrat       0.96      0.95      0.96        79
  republican       0.92      0.94      0.93        52

    accuracy                           0.95       131
   macro avg       0.94      0.95      0.94       131
weighted avg       0.95      0.95      0.95       131



### Машина опорных векторов (SVC)

In [274]:
from sklearn.svm import SVC

# Обьявляем
svm = SVC(kernel='linear', C=1.0, random_state=1)

# Обучение модели
svm.fit(x_train_std, y_train)

SVC(kernel='linear', random_state=1)

In [275]:
y_pred_svm = svm.predict(x_test_std)

In [276]:
# Выводим отчёт
print(classification_report(y_test, y_pred_svm, target_names=cvd_pred_names))

              precision    recall  f1-score   support

    democrat       0.96      0.96      0.96        79
  republican       0.94      0.94      0.94        52

    accuracy                           0.95       131
   macro avg       0.95      0.95      0.95       131
weighted avg       0.95      0.95      0.95       131



### Нелинейная классификация

In [277]:
svm_rbf = SVC(kernel='rbf', random_state=1, gamma=0.10, C=10.0)
svm_rbf.fit(x_train_std, y_train)

SVC(C=10.0, gamma=0.1, random_state=1)

In [278]:
y_pred_svm_rbf = svm_rbf.predict(x_test_std)

# Выводим отчёт
print(classification_report(y_test, y_pred_svm_rbf, target_names=cvd_pred_names))

              precision    recall  f1-score   support

    democrat       0.97      0.96      0.97        79
  republican       0.94      0.96      0.95        52

    accuracy                           0.96       131
   macro avg       0.96      0.96      0.96       131
weighted avg       0.96      0.96      0.96       131



### Дерево принятия решении

In [279]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier(criterion='gini',
                              max_depth=4,
                              random_state=1)
tree.fit(x_train_std, y_train)

DecisionTreeClassifier(max_depth=4, random_state=1)

In [280]:
y_pred_tree = tree.predict(x_test_std)
# Выводим отчёт
print(classification_report(y_test, y_pred_tree, target_names=cvd_pred_names))

              precision    recall  f1-score   support

    democrat       0.97      0.94      0.95        79
  republican       0.91      0.96      0.93        52

    accuracy                           0.95       131
   macro avg       0.94      0.95      0.94       131
weighted avg       0.95      0.95      0.95       131



### Метод ближайших соседей (kNN)

In [281]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors = 5, p=2, metric='manhattan')
knn.fit(x_train_std, y_train)

y_pred_knn = knn.predict(x_test_std)
# Выводим отчёт
print(classification_report(y_test, y_pred_knn, target_names=cvd_pred_names))

              precision    recall  f1-score   support

    democrat       0.96      0.89      0.92        79
  republican       0.84      0.94      0.89        52

    accuracy                           0.91       131
   macro avg       0.90      0.91      0.91       131
weighted avg       0.91      0.91      0.91       131



##  Вывод по датасету congressional_voting_dataset

### Самый лучший результат показывает алгоритм нелинейной классификации (SVC) с точностью в 0.96. Другие алгоритмы показывают результаты не сильно хуже:
2. 0.95 - машина опорных векторов, логистическая регрессия
3. 0.94 - дерево принятия решении
4. 0.90 - метод ближайших соседей

## 2. heart_disease_dataset

In [282]:
hdd = pd.read_csv("C:/Users/Даниил/Desktop/ОМО/heart_disease_dataset.csv")

# Получаем список значении
x = hdd.iloc[:, :-1].values

# Получаем список значении первичного параметра
y = hdd['num'].values

In [283]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
le.fit(y)
y = le.transform(y)

# Приводим названия в классы для отображения(?)
hdd_pred_names = le.classes_
hdd_pred_names

# Разделяем датасет на обучающие признаки, тестовые признаки, целевые обучающие, целевые тестовые
x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.3, random_state=0)

from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
sc.fit(x_train)
x_train_std = sc.transform(x_train)
x_test_std = sc.transform(x_test)

# Проверка перевода параметров в численный вид
#x_train[:5], x_train_std[:5]

### Логистическая регрессия

In [284]:
from sklearn.linear_model import LogisticRegression

# Создаем логистическую регрессию и обучаем ее (fit) 
lr = LogisticRegression(C=100.0, random_state = 1)
lr.fit(x_train_std, y_train)

LogisticRegression(C=100.0, random_state=1)

In [285]:
lr.predict_proba(x_test_std[:3, :])
# Выдает значения предсказания с наибольшей вероятностью
lr.predict_proba(x_test_std[:10, :]).argmax(axis=1)
# Делаем тестовое предсказание
lr.predict(x_test_std[0, :].reshape(1, -1)) 
x_test_std[0, :]
x_test_std[0, :].reshape(1, -1)

array([[-2.16925442, -1.50384124, -1.20851058, -0.83965571, -0.70790083,
        -0.40599897, -0.93609688,  1.88970222, -0.68718427, -0.30158238,
        -0.99027236,  0.06875015,  0.09739544]])

In [286]:
y_pred = lr.predict(x_test_std)
target_names=['No disease','Disease']
print(classification_report(y_test, y_pred, target_names=target_names))

              precision    recall  f1-score   support

  No disease       0.67      0.70      0.69        47
     Disease       0.67      0.64      0.65        44

    accuracy                           0.67        91
   macro avg       0.67      0.67      0.67        91
weighted avg       0.67      0.67      0.67        91



### Машина опорных векторов (SVC)

In [287]:
from sklearn.svm import SVC

# Обьявляем
svm = SVC(kernel='linear', C=1.0, random_state=1)

# Обучение модели
svm.fit(x_train_std, y_train)

y_pred_svm = svm.predict(x_test_std)

# Выводим отчёт
print(classification_report(y_test, y_pred_svm, target_names=target_names))

              precision    recall  f1-score   support

  No disease       0.68      0.77      0.72        47
     Disease       0.71      0.61      0.66        44

    accuracy                           0.69        91
   macro avg       0.69      0.69      0.69        91
weighted avg       0.69      0.69      0.69        91



### Нелинейная классификация

In [288]:
svm_rbf = SVC(kernel='rbf', random_state=1, gamma=0.10, C=10.0)
svm_rbf.fit(x_train_std, y_train)

y_pred_svm_rbf = svm_rbf.predict(x_test_std)

# Выводим отчёт
print(classification_report(y_test, y_pred_svm_rbf, target_names=target_names))

              precision    recall  f1-score   support

  No disease       0.74      0.72      0.73        47
     Disease       0.71      0.73      0.72        44

    accuracy                           0.73        91
   macro avg       0.73      0.73      0.73        91
weighted avg       0.73      0.73      0.73        91



### Дерево принятия решении

In [289]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier(criterion='gini',
                              max_depth=4,
                              random_state=1)
tree.fit(x_train_std, y_train)

y_pred_tree = tree.predict(x_test_std)
# Выводим отчёт
print(classification_report(y_test, y_pred_tree, target_names=target_names))

              precision    recall  f1-score   support

  No disease       0.75      0.85      0.80        47
     Disease       0.82      0.70      0.76        44

    accuracy                           0.78        91
   macro avg       0.79      0.78      0.78        91
weighted avg       0.78      0.78      0.78        91



### Метод ближайших соседей (kNN)

In [290]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors = 5, p=2, metric='manhattan')
knn.fit(x_train_std, y_train)

y_pred_knn = knn.predict(x_test_std)
# Выводим отчёт
print(classification_report(y_test, y_pred_knn, target_names=target_names))

              precision    recall  f1-score   support

  No disease       0.69      0.74      0.71        47
     Disease       0.70      0.64      0.67        44

    accuracy                           0.69        91
   macro avg       0.69      0.69      0.69        91
weighted avg       0.69      0.69      0.69        91



##  Вывод по датасету heart_disease_dataset


###  Самый лучший результат показывает алгоритм дерево принятия решении с точностью в 0.79. Другие алгоритмы показывают следующие результаты:
2. 0.73 - нелинейная классификация
3. 0.69 - машина опорных векторов, метод ближайших соседей
4. 0.67 - логистическая регрессия